### PR TITLE
Victory legend auto rows

### LINKED ISSUE #1197: VictoryLegend auto `itemsPerRow` based on desired width
# Feature Requests

### Checklist

* [Yes ] I've read through the [Docs](https://formidable.com/open-source/victory/docs) and [Guides](https://formidable.com/open-source/victory/guides) to make sure this functionality doesn't already exist

* [Yes ] I've searched open issues to make sure I'm not opening a duplicate issue

### Description

Often, I my intentions are less "I want a specific number of items per row/column in a legend" and more "I want as many items in a row/column as it takes to not overflow a certain width/height" (depending on horizontal/vertical orientation).

As the number and length of items are typically not known in advance, devs have to either make their own automatic behavior or force the end user to manually specify `itemsPerRow` with a textbox or something.

Implementation would likely require a new `legendWidth` prop to disambiguate from `width` that drives the container component. Probably a width/height-neutral term? But `itemsPerRow` already implicitly means `itemsPerColumn` when `orientation={'vertical'}` (which is the default?)? Eh, semantics. I'd acquiesce to whatever is deemed "least surprising".

I currently have a (rather naive) wrapper that solves the problem iteratively, without accounting for is-title-inline-with-desired-dimension, assuming desired dimension to be whole-svg `width` minus whole-svg left/right `padding`.

I think you basically *have* to work iteratively, since all the fields in a column share the (max) length. Thus, the maximum row length isn't trivial to calculate in closed form.

I think it would be possible to adapt the approach into `VictoryLegend` itself.
With sufficient guidance on high-level architecture concerns and on test design, I'd love to help contribute to this!

### PR DESCRIPTION
**Pull request by rudasoftware**
Fixes #1197 

`legendWidth` prop follows semantics established by `itemsPerRow`. 
  - Even though "width" and "row" are associated with "horizontal" orientation, they are used to mean "dimension that would be affected by wrapping in the direction specified by `orientation`". 
  - If there were also a `legendHeight` prop, one of the two props would be effectively meaningless depending on the orientation.

Behaves more like `legendMaxWidth`, I guess, because the `width` of `border` is not expanded to take up the full specified space. If that behavior is desired, it is theoretically possible to also hack `borderPadding`?

I tried my best to match existing code style, but it's probably not perfect...

### CODE REVIEW

**Comment by boygirl** [comment]
@rudasoftware thanks so much! First glance looks good. I'll do a thorough review over the weekend :)

**Comment by boygirl** [comment]
@rudasoftware This seems pretty close to mergeable to me. A couple things:

1) I agree that `maxLegendWidth` or maybe `targetLegendWidth` is a more descriptive prop name for how this operate. 

2) I can see a use case for `maxLegendHeight` as well. When a legend is split into rows the concept of horizontal / non-horizontal ends up referring  to the directional ordering than the overall dimension of the legend, and so you might still want to constrain a horizontal chart in the vertical dimension.

 For example: 
```jsx
<VictoryLegend
  orientation="horizontal"
  legendWidth={100}
  data={data}
/>
```
screen shot 2018-12-18 at 10 12 20 am


vs.
```
<VictoryLegend
  legendWidth={100}
  data={data}
/>
```
screen shot 2018-12-18 at 10 13 43 am

2) new props need to be added to `propTypes`

**Comment by rudasoftware** [comment]
Oops, yeah, `propTypes` is an obvious oversight.

**Comment by rudasoftware** [comment]
I think I understand what you want from `maxLegendWidth` and `maxLegendHeight`. I'm less sure about how to handle "overcontrained" situations where (due to styling, data count, and individual datum length) there is no possible way for the legend to fit in the box defined by both props together. Right now, if 'legendWidth' is too narrow, `VictoryLegend` essentially says, "well, I tried my best, but I really can't do more than `itemsPerRow={1}`, so that's what I'm giving you." Now what we want is an optimization problem of *two* variables, so the minimum case is less clear (at least to me).

Should they be mutually exclusive? (I tend to not like that.) Should one have precedence over the other, perhaps depending on `orientation`?

**Comment by rudasoftware** [comment]
I think it makes less sense to make it accept a function, unless we also exported some "preset" functions that would do the things. Functional styles make a lot of sense, because there are basically infinitely many customizations a user might want to make based on data. I think the problem space for this issue is much smaller. Just judging by the average Gitter question, I think this approach would be rather confusing for a lot of people. Even with very good documentation, I think writing a successful `itemsPerRow` function would be difficult for a user without a bit of knowledge about the `victory` codebase or previous experience with lower-level layout logic, somewhat defeating the purpose of a function prop as an "easy" avenue of extension.

Also, providing it with the first pass dimensions would be almost useless, I think, because the core of the algorithm is iterative/recursive; the function would need to be called multiple times with updated dimensions each time... or maybe (to support looping) receive a callable that returns the dimensions given current props? But then you'd have to deal with wrangling the props from all appropriate sources and make sure they're in sync and correctly overriding. It sounds like a good idea on the surface, but I think it quickly balloons into complexity for both the user and the implementation.

TL;DR I think that an `itemsPerRow` function prop wouldn't really improve usability. Everyone who wanted this functionality would more or less have to rewrite most of what my wrapper did in a slightly different mechanism.

**Comment by rudasoftware** [comment]
Maybe it is just better to have `maxLegendWidth` and `maxLegendHeight` and one just "wins" over the other depending on `orientation`?

E.g. You *can* `maxLegendHeight` on a horizontal chart, but if you add `maxLegendWidth` it takes over.

Maybe back up documented behavior with a `propTypes` warning?

I thought about a "scoring" algorithm, where each iteration keeps track of the error between max and actual size for both dimensions and then picks the iteration where both scores are closest to 0... but I'm pretty sure that in most cases this winds up looking the same as ignoring the subordinate dimension. In "over constrained" cases, though, the resulting dimensions that doesn't fully obey either max would probably still be more confusing (and unpredictable) rather than empowering for the user.

**Comment by boygirl** [comment]
Yes, I think having one of the constraints "win" is the least confusing of the possible implementations, and a `Log.warn` when both props are given is definitely desirable.

**Comment by bgold0** [comment]
Looking forward to this being pushed! Any ETA?

**Comment by rudasoftware** [comment]
Heh, sorry @bgold0, I got sidetracked on to other projects after my own custom implementation reached "good enough" status. When I get a free moment one of these days, I'd like to come back to this and polish it up.

**Comment by boygirl** [comment]
@rudasoftware funny timing. I'm planning on some breaking changes for `VictoryLegend` in `victory@33.0.0`, and targeting the end of the month for the release. If you decide to pick this back up, drop me a note so we can coordinate.

**Comment by pramodgupta** [comment]
@rudasoftware I'm interested in this development as we could use some of this dev.. Do you have any timeline on when this could be in?

**Comment by Fernando555** [comment]
is there any progress on this, really really need this feature D:
